Add new endpoint and handler for synchronous prediction requests

acrellin opened the pull request.

acrellin committed:
Add route handler for simple, synchronous prediction of single time series

acrellin committed:
Add test for /predict_single handler

bnaul commented on cesium_app/handlers/prediction.py:
```diff
@@ -5,11 +5,13 @@
 
 import tornado.gen
 from tornado.web import RequestHandler
+import tornado.escape
```
Maybe just import `json_decode` directly so those calls will be shorter?

bnaul commented on cesium_app/handlers/prediction.py:
```diff
@@ -137,3 +139,28 @@ def delete(self, prediction_id):
+    def post(self):
+        ts_data = tornado.escape.json_decode(self.get_argument('ts_data'))
+        model_id = int(self.get_argument('modelID'))
+        features_to_use = tornado.escape.json_decode(
```
Aren't the features to use stored in the model? The user shouldn't be able to choose this.

bnaul commented on cesium_app/handlers/prediction.py:
```diff
@@ -137,3 +139,28 @@ def delete(self, prediction_id):
+
+        fset_data = cesium.featurize.featurize_time_series(
+            *ts_data, features_to_use=features_to_use, meta_features=meta_feats)
+        fset = cesium.featureset.Featureset(fset_data).impute()
```
Should imputation strategy be an option that the user can POST? I can't even remember what the default is actually...

stefanv commented:
I just chatted with Josh about this, and he says it would be nice if we have the ability to persist the result (i.e., automatically go through the steps of "data -> featurize -> model -> predict".

stefanv commented:
Also, I guess we want to use the features for the model we're predicting on, not all of them?

acrellin commented:
@stefanv can you clarify what you mean by automatically go through the steps?

acrellin commented:
@stefanv it's not using all of them, only if "all" is specified (as in the test)

stefanv commented:
Just persist the resulting dataset, featureset, model, and prediction.

acrellin commented:
@stefanv so scrapping this approach altogether?

acrellin commented:
@bnaul I wanted to leave room for the case where a model may exist, but doesn't have an associated `featureset` (which the approach you suggested assumes it has), e.g. if one creates the DB state manually with just a pre-built model

acrellin committed:
Allow imputation strategy & value to be passed in as query parameters
Get features list from featureset associated with model
Rename predict_single to predict_raw_data

acrellin commented:
@bnaul implemented your other two suggestions, though. thanks for the feedback

bnaul commented:
Hm...that still makes me a bit uneasy because you could get into a situation where a model was trained on `['amplitude', 'minimum']` but then for new data we're feeding it `['amplitude', 'maximum']` and no error is raised.

acrellin commented:
@bnaul Yeah, I see your point. The two routes I see are 1) requiring that the `model` have an associated `featureset`, or 2) adding an optional `features_list` attribute to the `Model` model, so this will still work in the case I described above.

bnaul commented:
The case you described is impossible right now, though; were we going to add support for that for some reason?

```
Foreign-key constraints:
    "model_featureset_id_fkey" FOREIGN KEY (featureset_id) REFERENCES featureset(id) ON DELETE CASCADE
```

acrellin commented:
I had been thinking about it, but think it's fine to go with option 1) for now. I'll update this PR to accord with that.

stefanv commented on cesium_app/app_server.py:
```diff
@@ -32,6 +33,7 @@ def make_app():
         (r'/models(/.*)?', ModelHandler),
         (r'/predictions(/[0-9]+)?', PredictionHandler),
         (r'/predictions/([0-9]+)/(download)', PredictionHandler),
+        (r'/predict_raw_data(/.*)?', PredictRawDataHandler),
```
remove the `(/.*)?`?

stefanv commented on cesium_app/tests/frontend/test_predict_raw_data.py:
```diff
@@ -0,0 +1,19 @@
+         create_test_featureset(p) as fs, create_test_model(fs) as m:
+        ts_data = [[1, 2, 3, 4], [32.2, 53.3, 32.3, 32.52], [0.2, 0.3, 0.6, 0.3]]
+        impute_kwargs = {'strategy': 'constant', 'value': None}
+        query_string = ('http://localhost:5000/predict_raw_data?ts_data={}'
```
Let's not hardcode things like ports---I use a different port to test on my machine.

stefanv replied:
Also, let's not do a query but handle this as a POST.

acrellin committed:
Get base url from config
